taichi/program/state_flow_graph.h

```diff
@@ -21,9 +22,121 @@ class AsyncEngine;
 class StateFlowGraph {
  public:
   struct Node;
-  using StateToNodesMap =
-      llvm::SmallVector<std::pair<AsyncState, llvm::SmallSet<Node *, 8>>, 4>;
 
+  // A specialized container for fast edge insertion and lookup
+  class StateToNodesMap {
+   private:
+    using MaskType = uint64;  // Unused yet
+    static constexpr unsigned kNumInlined = sizeof(MaskType) * 8u;
```

Comment:
I believe the small vector inlined size should be *greater than or equal to* the number of bits in the mask:

* Inside `StateToNodesMap`'s constructor, we need to resize `data_` to #bits, so that if an edge maps to a slot `i` in the mask, we can directly call `data_[i] = edge;`. 
* If `kNumInlined` is smaller than #bits of the mask, then `resize()` will exceed the inlined region size. So small vector effectively becomes `std::vector`.
* But as `kNumInlined` becomes larger, it wastes more space. Right now for a `uint64` mask, `kNumInlined` need to be >= 64. We might need to tweak this between 32 and 64 to strike a balance..